### ISSUE TITLE
Combine finite-model building with refutation search

### ISSUE THREAD

**Issue by Columpio** [vprover/vampire#274]
Finite model building halts with SIGSEGV on the following **unsatisfiable** problem:
[input.smt2](https://github.com/vprover/vampire/files/7225154/iteration_0.FreeSorts.txt)

My idea is that `fmb` should build refutations for unsatisfiable problems as well as models for satisfiable ones.

(Or there should be such portfolio mode)

# Details

Usual run of Vampire without `-sa fmb` options results in refutation.

Calling `vampire --input_syntax smtlib2 --saturation_algorithm fmb input.smt2` results with the following.
#### On release (`Vampire 4.5.1 (commit 57a6f78c on 2020-07-15 11:59:04 +0200)`)
```
TRYING [1]
TRYING [2]
48862 Aborted by signal SIGSEGV on /tmp/1632471437692372446/iteration_0.FreeSorts.smt2
% ------------------------------
% Version: Vampire 4.5.1 (commit 57a6f78c on 2020-07-15 11:59:04 +0200)
% Termination reason: Unknown
% Termination phase: Finite model building constraint generation

% Memory used [KB]: 5628
% Time elapsed: 0.008 s
% ------------------------------
% ------------------------------
```
#### On `induction-recursive-functions2` branch (with recent fixes) (`Vampire 4.5.1 (commit 29a913ad2 on 2021-09-15 16:02:43 +0200)`)

```
TRYING [1]
TRYING [2]
Condition vvar <= (unsigned)_solver.nVars() in file /home/vampire/SAT/MinisatInterfacingNewSimp.hpp, line 120 was violated, as:
vvar == 262291
(unsigned)_solver.nVars() == 576
----- stack dump -----
Version : Vampire 4.5.1 (commit 29a913ad2 on 2021-09-15 16:02:43 +0200)
Control points passed: 1523779
last control point:
         +MinisatInterfacingNewSimp::addClause (1523779)
main
 vampireMode()
  doProving()
   ProvingHelper::runVampireSaturation
    ProvingHelper::runVampireSaturationImpl
     MainLoop::run
      FiniteModelBuilder::runImpl
       SATSolver::addClauses
        MinisatInterfacingNewSimp::addClause
----- end of stack dump -----
```
#### On `induction-recursive-functions` branch (`Vampire 4.5.1 (commit 00e960980 on 2021-08-04 15:31:24 +0200)`)
(the same as `induction-recursive-functions2`)

**Comment by selig** [comment]
Thanks, I'll take a look into this issue. The assertion violation is saying that the number of variables created in the SAT solver is different from the number being created in the propositional translation - so clearly something is wrong.

A couple of notes.

Most of the portfolio modes will try both saturation and fmb e.g. running the smtcomp portfolio on a UF problem. But these portfolios will also sometimes return saturations for a satisfiable problem or they might assert unsatisfiability due to the lack of a finite model of a certain size required by a domain axiom. Do you want a portfolio that only either returns finite models or derivations?

There are also known issues with finite model building and this may be related (although the assertion violation isn't the same). There is a need to perform a major refactoring of the preprocessing pipeline to avoid the many workarounds in finite model building relate to changes to the signature during preprocessing. It's on the backlog as nobody is currently available to do this.

**Comment by Columpio** [comment]
> Do you want a portfolio that only either returns finite models or derivations?

Yes, exactly. The only alternative for me is CVC's finite-model-finding option, which gives both `sat` and `unsat` answers. However, CVC does not return a refutation, and the developers do not have any plans to implement that.

That's why I will be glad to use Vampire for such a task.
> There are known issues with fmb.

I know some of them (as I reported them), that's why I use `induction-recursive-functions` branch, which has some fixes for `fmb`. Could you please tell me whether there is a more stable branch with respect to finite-model building?

**Comment by quickbeam123** [comment]
Hi Yurii,

this issue is now fixed with 916b28ab14fefcfaff in master. 

But, just to be clear, I mean the assertion violation issue, not what the title asks for! (The assertion violation did not stem from the problem being unsat... Our finite model building can sometimes report UNSAT, but only very rarely is FMB the strategy to use to show a formula unsatisfiable - you are almost always better off trying some other strategies instead!)

**Comment by Columpio** [comment]
> Do you want a portfolio that only either returns finite models or derivations?

Yes, exactly. The only alternative for me is CVC's finite-model-finding option, which gives both `sat` and `unsat` answers. However, CVC does not return a refutation, and the developers do not have any plans to implement that.

That's why I will be glad to use Vampire for such a task.
> There are known issues with fmb.

I know some of them (as I reported them), that's why I use `induction-recursive-functions` branch, which has some fixes for `fmb`. Could you please tell me whether there is a more stable branch with respect to finite-model building?
